sql: index out of range error in queries with set operations

roachtest.sqlsmith/setup=seed/setting=no-mutations [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2920165&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2920165&tab=artifacts#/sqlsmith/setup=seed/setting=no-mutations) on master @ [3cd29547884feb24b150b67ca2a88073798a35fd](https://github.com/cockroachdb/cockroach/commits/3cd29547884feb24b150b67ca2a88073798a35fd):


```
								(e'!\\f\t\x1b#':::STRING:::NAME),
								(e'\n\x1a':::STRING:::NAME),
								(e's\x05\x0f&\x11Ct':::STRING:::NAME),
								(
									CASE
									WHEN ('102.216.247.91/23':::INET::INET >> '193.136.21.129/16':::INET::INET)::BOOL
									THEN 'W/07Vu':::STRING:::NAME
									ELSE e'\x1b4\'r!\t':::STRING:::NAME
									END
								)
						)
							AS tab_28 (col_41)
				)
		SELECT
			NULL AS col_42
		FROM
			with_5 AS cte_ref_2
		WHERE
			true
		ORDER BY
			cte_ref_2.col_39
		LIMIT
			78:::INT8;

	cluster.go:1716,context.go:140,cluster.go:1705,test_runner.go:858: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-2920165-1619291008-15-n4cpu4 --oneshot --ignore-empty-nodes: exit status 1 1: dead
		3: 13772
		2: 15100
		4: 12381
		Error: UNCLASSIFIED_PROBLEM: 1: dead
		(1) UNCLASSIFIED_PROBLEM
		Wraps: (2) attached stack trace
		  -- stack trace:
		  | main.glob..func14
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1162
		  | main.wrap.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:284
		  | github.com/spf13/cobra.(*Command).execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:856
		  | github.com/spf13/cobra.(*Command).ExecuteC
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:960
		  | github.com/spf13/cobra.(*Command).Execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:897
		  | main.main
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:2057
		  | runtime.main
		  | 	/usr/local/go/src/runtime/proc.go:204
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1374
		Wraps: (3) 1: dead
		Error types: (1) errors.Unclassified (2) *withstack.withStack (3) *errutil.leafError
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
# From https://go.crdb.dev/p/roachstress, perhaps edited lightly.
caffeinate ./roachstress.sh sqlsmith/setup=seed/setting=no-mutations
```
</p>
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=seed/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

First instance is:
```
panic: runtime error: index out of range [1] with length 1 [recovered]
	panic: runtime error: index out of range [1] with length 1

goroutine 2562 [running]:
panic(0x486a0a0, 0xc002d6fad0)
	/usr/local/go/src/runtime/panic.go:1064 +0x545 fp=0xc0016934e0 sp=0xc001693418 pc=0x48b725
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc0017a8c00, 0x5b6ba20, 0xc0022c09c0, 0x4a40000, 0xc002358240)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:814 +0x3a9 fp=0xc001693560 sp=0xc0016934e0 pc=0x3537ac9
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc0017a8c00, 0x5b6ba20, 0xc0022c09c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:482 +0x65 fp=0xc001693598 sp=0xc001693560 pc=0x370e4a5
runtime.call32(0x0, 0x53eddd8, 0xc0015734e8, 0x1800000018)
	/usr/local/go/src/runtime/asm_amd64.s:540 +0x3e fp=0xc0016935c8 sp=0xc001693598 pc=0x4c30be
runtime.reflectcallSave(0xc001693708, 0x53eddd8, 0xc0015734e8, 0x18)
	/usr/local/go/src/runtime/panic.go:881 +0x58 fp=0xc0016935f8 sp=0xc0016935c8 pc=0x48b118
runtime.runOpenDeferFrame(0xc0006af680, 0xc0015734a0, 0x0)
	/usr/local/go/src/runtime/panic.go:855 +0x2cd fp=0xc001693688 sp=0xc0016935f8 pc=0x48afcd
panic(0x4a40000, 0xc002358240)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9 fp=0xc001693750 sp=0xc001693688 pc=0x48b399
runtime.goPanicIndex(0x1, 0x1)
	/usr/local/go/src/runtime/panic.go:88 +0xa5 fp=0xc001693798 sp=0xc001693750 pc=0x489225
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).bufferRow(0xc0014f4900, 0x5b89f20, 0xc000d77838, 0xc00285c8a0, 0x2, 0x2, 0x0, 0x0, 0x0, 0x200000000, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:1238 +0x3f0 fp=0xc001693820 sp=0xc001693798 pc=0x3bfcf70
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*commandResult).AddRow(0xc0014f4f80, 0x5b89f20, 0xc000d77838, 0xc00285c8a0, 0x2, 0x2, 0x2, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/command_result.go:206 +0x19b fp=0xc0016938c0 sp=0xc001693820 pc=0x3bf57db
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLReceiver).Push(0xc001635c00, 0xc002cccf00, 0x2, 0x2, 0x0, 0xc001c00318)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:740 +0x8f6 fp=0xc001693a20 sp=0xc0016938c0 pc=0x35bca76
github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run(0x5b89f20, 0xc001c00318, 0x5b908a0, 0xc001be8600, 0x5b33ce0, 0xc001635c00)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:178 +0x5a fp=0xc001693a60 sp=0xc001693a20 pc=0x245c07a
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).Run(0xc001be8600, 0x5b6ba20, 0xc001f90300)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:803 +0x96 fp=0xc001693aa0 sp=0xc001693a60 pc=0x2463496
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run(0xc000f4ca20, 0x5b6ba20, 0xc001f90300, 0x53edd68, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:384 +0x2e2 fp=0xc001693b48 sp=0xc001693aa0 pc=0x26dd2c2
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc000a23130, 0xc0013a5a40, 0xc001c61200, 0xc00321b200, 0xc001635c00, 0xc0017a92f8, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:357 +0x472 fp=0xc001694090 sp=0xc001693b48 pc=0x35bad12
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc000a23130, 0x5b89f20, 0xc000d77838, 0xc0017a92f8, 0xc0013a5a40, 0xc001c61200, 0x5b6dae0, 0xc0010e6fa0, 0x0, 0xc001635c00, ...)
...
```

--- a/pkg/sql/logictest/testdata/logic_test/union
+++ b/pkg/sql/logictest/testdata/logic_test/union
@@ -543,3 +543,19 @@ SELECT a, b FROM ab EXCEPT ALL SELECT x, y FROM xy ORDER BY a
 ----
 2  2
 4  4
+
+# Regression test for #64181. Ensure that a projection on top of an ordered
+# UNION ALL correctly projects away ordering columns.
+query TT
+WITH q (x, y) AS (
+  SELECT * FROM (VALUES ('a', 'a'), ('b', 'b'), ('c', 'c'))
+  UNION ALL
+  SELECT * FROM (VALUES ('d', 'd'))
+)
+SELECT 'e', y FROM q
+ORDER BY x
+----
+e  a
+e  b
+e  c
+e  d

--- a/pkg/sql/opt/exec/execbuilder/testdata/union
+++ b/pkg/sql/opt/exec/execbuilder/testdata/union
@@ -755,3 +755,57 @@ vectorized: true
               spans: FULL SCAN
 ·
 Diagram: https://cockroachdb.github.io/distsqlplan/decode.html#eJyskt-Lm04Uxd-_f8XlPuk3U7L-WspAYLKNywpp3Kr0ByUPRm-zgqt2ZgJdQv73ohZ2TRPbtH2cM_O555yre1RfS-Tof7xfzoMVGIsgTuJ3Swbv_egmjH0TYn_pv0ngf7iNwrdgDI_pJssJglXiR3Erz5dLOPHChA93fuSDkcEMLBPmqwUYOcyATAijhR_BzSfYMMgY5AyIQYoMqzqnVfpICvlntHDNsJF1RkrVspX23YMg_4b8imFRNTvdymuGWS0J-R51oUtCjkm6KSmiNCc5vUKGOem0KLuxXTrRyOIxlU_IMG7SSnF4hesDw3qnn4cqnW4JuXVgv298W5SaJMmpNXTtdQ6GcNp9cM6DVfL6x1qECzMQnnk2gn1JhJfd7b_s7vxRd-dfdnfPRnh23lW1zElSPjBet-SvnpzocZeqh5h02EzdYZHkqSE-_PWRYUlftCEsJmwmHCZcJjxzJovtw08qMgx3msNQPtvcu2T5cS01yak3TCzsCRPOhAl3woQ1OWt1fYlVRKqpK0XHyz45-ardMOVb6r-Yqncyo3tZZ51Nfww7rhNyUrq_tfpDUPVXbcCXsDUKuwPYPobtUdgZd3YucLaOYXcU9sadvVH4-gheH_77HgAA__-rHNsg
+
+# Regression test for #64181. Ensure that a projection on top of an ordered
+# UNION ALL correctly projects away ordering columns.
+query T
+EXPLAIN (DISTSQL,VERBOSE) WITH q (x, y) AS (
+  SELECT * FROM (VALUES ('a', 'a'), ('b', 'b'), ('c', 'c'))
+  UNION ALL
+  SELECT * FROM (VALUES ('d', 'd'))
+)
+SELECT 'e', y FROM q
+ORDER BY x
+----
+distribution: local
+vectorized: true
+·
+• project
+│ columns: ("?column?", y)
+│
+└── • render
+    │ columns: ("?column?", x, y)
+    │ ordering: +x
+    │ estimated row count: 4
+    │ render ?column?: 'e'
+    │ render x: column1
+    │ render y: column2
+    │
+    └── • union all
+        │ columns: (column1, column2)
+        │ ordering: +column1
+        │ estimated row count: 4
+        │
+        ├── • sort
+        │   │ columns: (column1, column2)
+        │   │ ordering: +column1
+        │   │ estimated row count: 3
+        │   │ order: +column1
+        │   │
+        │   └── • values
+        │         columns: (column1, column2)
+        │         size: 2 columns, 3 rows
+        │         row 0, expr 0: 'a'
+        │         row 0, expr 1: 'a'
+        │         row 1, expr 0: 'b'
+        │         row 1, expr 1: 'b'
+        │         row 2, expr 0: 'c'
+        │         row 2, expr 1: 'c'
+        │
+        └── • values
+              columns: (column1, column2)
+              size: 2 columns, 1 row
+              row 0, expr 0: 'd'
+              row 0, expr 1: 'd'
+·
+Diagram: https://cockroachdb.github.io/distsqlplan/decode.html#eJykkUFv1DAQhe_8itFcHIPRJllOPnVLA0QKSYm3BYT2kMajaqUQZ20HbbXa_46SCJWtmqqF23j8vvfGngO6XYMSk2-X2SrNIbhI1Vp9yQRcJ-V5oRIOX9P1J9hBsBdwx2GlIFBJlrxfw2v4UBafIbheZVeJgoBVTACrGBcQsJuhvpnqeqhrxjlc5WmRwyrLYM5DD1rNOOd_JIyYgLtJt4OivEhKOP8OexTYGk159ZMcyh8Y4UZgZ01Nzhk7tA6jINV7lKHAbdv1fmhvBNbGEsoD-q1vCCU2pq4a-FU1PTkIFyEK1OSrbTPqjwJN7-9p56tbQhkdxT8mRIvoWQnxSxKUsZ7sIj5xxrPoDc7ZL2ft712N1WRJP2b6yAy5eWu6xfJUXVKrycphkVJKtS7T_KOAs3h2rvAlzy7JdaZ19KwPDYehSd_S9EhnelvTpTX1GDMdi5EbG5qcn26X0yFtx6tx83_D0ZNwfAKHD-H4f5KXT8LvHiRvjq9-BwAA__9bjjq0

--- a/pkg/sql/physicalplan/physical_plan.go
+++ b/pkg/sql/physicalplan/physical_plan.go
@@ -428,6 +428,10 @@ func (p *PhysicalPlan) EnsureSingleStreamOnGateway() {
 			panic("ensuring a single stream on the gateway failed")
 		}
 	}
+	// We now must have a single stream in the whole physical plan, so there is no
+	// ordering to be maintained for the merge of multiple streams. This also
+	// adheres to the comment on p.MergeOrdering.
+	p.MergeOrdering = execinfrapb.Ordering{}
 }
 
 // CheckLastStagePost checks that the processors of the last stage of the

